hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/chillmode/ChillModeExitRule.java

```diff
@@ -17,16 +17,87 @@
  */
 package org.apache.hadoop.hdds.scm.chillmode;
 
+import org.apache.hadoop.hdds.server.events.EventHandler;
+import org.apache.hadoop.hdds.server.events.EventPublisher;
+
+
 /**
- * Interface for defining chill mode exit rules.
+ * Abstract class for ChillModeExitRules. When a new rule is added, the new
+ * rule should extend this abstract class.
+ *
+ * Each rule Should do:
+ * 1. Should add a handler for the event it is looking for during the
+ * initialization of the rule.
+ * 2. Add the rule in ScmChillModeManager to list of the rules.
+ *
  *
  * @param <T>
  */
-public interface ChillModeExitRule<T> {
+public abstract class ChillModeExitRule<T> implements EventHandler<T> {
+
+  private final SCMChillModeManager chillModeManager;
+  private final String ruleName;
+
+  public ChillModeExitRule(SCMChillModeManager chillModeManager,
+      String ruleName) {
+    this.chillModeManager = chillModeManager;
+    this.ruleName = ruleName;
+  }
+
+  /**
+   * Return's the name of this ChillModeExit Rule.
+   * @return ruleName
+   */
+  public String getRuleName() {
+    return ruleName;
+  }
+
+
+  /**
+   * Validate's this rule. If this rule condition is met, returns true, else
+   * returns false.
+   * @return boolean
+   */
+  public abstract boolean validate();
+
+  /**
+   * Actual processing logic for this rule.
+   * @param report
+   */
+  public abstract void process(T report);
+
+  /**
+   * Cleanup action's need to be done, once this rule is satisfied.
+   */
+  public abstract void cleanup();
+
+  @Override
+  public final void onMessage(T report, EventPublisher publisher) {
+
+    // TODO: when we have remove handlers, we can remove getInChillmode check
+
+    if (chillModeManager.getInChillMode()) {
+      if (validate()) {
+        cleanup();
+        chillModeManager.validateChillModeExitRules(ruleName, publisher);
+        return;
+      }
+
+      process(report);
 
-  boolean validate();
+      if (validate()) {
+        cleanup();
+        chillModeManager.validateChillModeExitRules(ruleName, publisher);
+      }
+    }
+  }
 
-  void process(T report);
+  /**
+   * Return SCMChillModeManager.
+   * @return SCMChillModeManager
+   */
+  public SCMChillModeManager getChillModeManager() {
+    return chillModeManager;
+  }
```

Comment:
We can replace this with 
`protected boolean scmInChillMode() {
	return chillModeManager..getInChillMode();
}`

Reply:
Done